File: medbrains/crates/medbrains-core/src/infection_control.rs
use chrono::{DateTime, NaiveDate, Utc};
use rust_decimal::Decimal;
use serde::{Deserialize, Serialize};
use serde_json::Value;
use uuid::Uuid;

// -- Enums -------------------------------------------------------------------

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, sqlx::Type)]
#[sqlx(type_name = "hai_type", rename_all = "snake_case")]
#[serde(rename_all = "snake_case")]
pub enum HaiType {
    Clabsi,
    Cauti,
    Vap,
    Ssi,
    Cdiff,
    Mrsa,
    Other,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, sqlx::Type)]
#[sqlx(type_name = "infection_status", rename_all = "snake_case")]
#[serde(rename_all = "snake_case")]
pub enum InfectionStatus {
    Suspected,
    Confirmed,
    RuledOut,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, sqlx::Type)]
#[sqlx(type_name = "antibiotic_request_status", rename_all = "snake_case")]
#[serde(rename_all = "snake_case")]
pub enum AntibioticRequestStatus {
    Pending,
    Approved,
    Denied,
    Expired,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, sqlx::Type)]
#[sqlx(type_name = "antibiotic_action", rename_all = "snake_case")]
#[serde(rename_all = "snake_case")]
pub enum AntibioticAction {
    Initiate,
    Escalate,
    DeEscalate,
    Stop,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, sqlx::Type)]
#[sqlx(type_name = "waste_category", rename_all = "snake_case")]
#[serde(rename_all = "snake_case")]
pub enum WasteCategory {
    Yellow,
    Red,
    WhiteTranslucent,
    Blue,
    Cytotoxic,
    Chemical,
    Radioactive,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, sqlx::Type)]
#[sqlx(type_name = "outbreak_status", rename_all = "snake_case")]
#[serde(rename_all = "snake_case")]
pub enum OutbreakStatus {
    Suspected,
    Confirmed,
    Contained,
    Closed,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, sqlx::Type)]
#[sqlx(type_name = "hygiene_moment", rename_all = "snake_case")]
#[serde(rename_all = "snake_case")]
pub enum HygieneMoment {
    BeforePatient,
    BeforeAseptic,
    AfterBodyFluid,
    AfterPatient,
    AfterSurroundings,
}

// -- Structs -----------------------------------------------------------------

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
pub struct InfectionSurveillanceEvent {
    pub id: Uuid,
    pub tenant_id: Uuid,
    pub patient_id: Uuid,
    pub admission_id: Option<Uuid>,
    pub hai_type: HaiType,
    pub infection_status: InfectionStatus,
    pub organism: Option<String>,
    pub susceptibility_pattern: Option<Value>,
    pub device_type: Option<String>,
    pub insertion_date: Option<DateTime<Utc>>,
    pub infection_date: DateTime<Utc>,
    pub location_id: Option<Uuid>,
    pub department_id: Option<Uuid>,
    pub nhsn_criteria: Option<String>,
    pub contributing_factors: Option<Value>,
    pub notes: Option<String>,
    pub reported_by: Uuid,
    pub confirmed_by: Option<Uuid>,
    pub confirmed_at: Option<DateTime<Utc>>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
pub struct InfectionDeviceDay {
    pub id: Uuid,
    pub tenant_id: Uuid,
    pub location_id: Uuid,
    pub department_id: Option<Uuid>,
    pub record_date: NaiveDate,
    pub patient_days: i32,
    pub central_line_days: i32,
    pub urinary_catheter_days: i32,
    pub ventilator_days: i32,
    pub recorded_by: Uuid,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
pub struct AntibioticStewardshipRequest {
    pub id: Uuid,
    pub tenant_id: Uuid,
    pub patient_id: Uuid,
    pub antibiotic_name: String,
    pub dose: Option<String>,
    pub route: Option<String>,
    pub frequency: Option<String>,
    pub duration_days: Option<i32>,
    pub indication: String,
    pub culture_sent: bool,
    pub culture_result: Option<String>,
    pub request_status: AntibioticRequestStatus,
    pub requested_by: Uuid,
    pub requested_at: DateTime<Utc>,
    pub reviewed_by: Option<Uuid>,
    pub reviewed_at: Option<DateTime<Utc>>,
    pub review_notes: Option<String>,
    pub escalation_reason: Option<String>,
    pub auto_stop_date: Option<NaiveDate>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
pub struct AntibioticConsumptionRecord {
    pub id: Uuid,
    pub tenant_id: Uuid,
    pub department_id: Option<Uuid>,
    pub antibiotic_name: String,
    pub atc_code: Option<String>,
    pub record_month: NaiveDate,
    pub quantity_used: Decimal,
    pub ddd: Option<Decimal>,
    pub patient_days: i32,
    pub ddd_per_1000_patient_days: Option<Decimal>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
pub struct BiowasteRecord {
    pub id: Uuid,
    pub tenant_id: Uuid,
    pub department_id: Uuid,
    pub waste_category: WasteCategory,
    pub weight_kg: Decimal,
    pub record_date: NaiveDate,
    pub container_count: i32,
    pub disposal_vendor: Option<String>,
    pub manifest_number: Option<String>,
    pub notes: Option<String>,
    pub recorded_by: Uuid,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
pub struct NeedleStickIncident {
    pub id: Uuid,
    pub tenant_id: Uuid,
    pub incident_number: String,
    pub staff_id: Uuid,
    pub incident_date: DateTime<Utc>,
    pub location_id: Option<Uuid>,
    pub department_id: Option<Uuid>,
    pub device_type: String,
    pub procedure_during: Option<String>,
    pub body_part: Option<String>,
    pub depth: Option<String>,
    pub source_patient_id: Option<Uuid>,
    pub hiv_status: Option<String>,
    pub hbv_status: Option<String>,
    pub hcv_status: Option<String>,
    pub pep_initiated: bool,
    pub pep_details: Option<String>,
    pub follow_up_schedule: Option<Value>,
    pub outcome: Option<String>,
    pub reported_by: Uuid,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
pub struct HandHygieneAudit {
    pub id: Uuid,
    pub tenant_id: Uuid,
    pub audit_date: DateTime<Utc>,
    pub location_id: Option<Uuid>,
    pub department_id: Uuid,
    pub auditor_id: Uuid,
    pub observations: i32,
    pub compliant: i32,
    pub non_compliant: i32,
    pub compliance_rate: Option<Decimal>,
    pub moment_breakdown: Option<Value>,
    pub staff_category: Option<String>,
    pub findings: Option<String>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
pub struct CultureSurveillance {
    pub id: Uuid,
    pub tenant_id: Uuid,
    pub culture_type: String,
    pub sample_site: String,
    pub location_id: Option<Uuid>,
    pub department_id: Option<Uuid>,
    pub collection_date: DateTime<Utc>,
    pub result: Option<String>,
    pub organism: Option<String>,
    pub colony_count: Option<i32>,
    pub acceptable: Option<bool>,
    pub action_taken: Option<String>,
    pub collected_by: Uuid,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
pub struct OutbreakEvent {
    pub id: Uuid,
    pub tenant_id: Uuid,
    pub outbreak_number: String,
    pub organism: String,
    pub outbreak_status: OutbreakStatus,
    pub detected_date: DateTime<Utc>,
    pub location_id: Option<Uuid>,
    pub department_id: Option<Uuid>,
    pub initial_cases: i32,
    pub total_cases: i32,
    pub description: Option<String>,
    pub control_measures: Option<Value>,
    pub hicc_notified: bool,
    pub hicc_notified_at: Option<DateTime<Utc>>,
    pub containment_date: Option<DateTime<Utc>>,
    pub closure_date: Option<DateTime<Utc>>,
    pub root_cause: Option<String>,
    pub lessons_learned: Option<String>,
    pub reported_by: Uuid,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
pub struct OutbreakContact {
    pub id: Uuid,
    pub tenant_id: Uuid,
    pub outbreak_id: Uuid,
    pub patient_id: Option<Uuid>,
    pub staff_id: Option<Uuid>,
    pub contact_type: String,
    pub exposure_date: Option<DateTime<Utc>>,
    pub screening_date: Option<DateTime<Utc>>,
    pub screening_result: Option<String>,
    pub quarantine_required: bool,
    pub quarantine_start: Option<NaiveDate>,
    pub quarantine_end: Option<NaiveDate>,
    pub notes: Option<String>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

File: medbrains/packages/types/src/infection-control.ts
// Infection control types — split from index.ts, barrel-re-exported.

// ── Infection Control ─────────────────────────────────────
export type HaiType = "clabsi" | "cauti" | "vap" | "ssi" | "cdiff" | "mrsa" | "other";
export type InfectionStatusType = "suspected" | "confirmed" | "ruled_out";
export type AntibioticRequestStatusType = "pending" | "approved" | "denied" | "expired";
export type WasteCategoryType =
  | "yellow"
  | "red"
  | "white_translucent"
  | "blue"
  | "cytotoxic"
  | "chemical"
  | "radioactive";
export type OutbreakStatusType = "suspected" | "confirmed" | "contained" | "closed";

export interface InfectionSurveillanceEvent {
  id: string;
  tenant_id: string;
  patient_id: string;
  admission_id?: string;
  hai_type: HaiType;
  infection_status: InfectionStatusType;
  organism?: string;
  susceptibility_pattern?: unknown;
  device_type?: string;
  insertion_date?: string;
  infection_date: string;
  location_id?: string;
  department_id?: string;
  nhsn_criteria?: string;
  contributing_factors?: unknown;
  notes?: string;
  reported_by: string;
  confirmed_by?: string;
  confirmed_at?: string;
  created_at: string;
  updated_at: string;
}

export type IndwellingDeviceType =
  | "central_line"
  | "urinary_catheter"
  | "ventilator"
  | "peripheral_iv"
  | "other";

export interface IndwellingDevice {
  id: string;
  tenant_id: string;
  patient_id: string;
  admission_id: string | null;
  device_type: IndwellingDeviceType;
  site: string | null;
  indication: string;
  inserted_at: string;
  inserted_by: string | null;
  last_reviewed_at: string;
  still_indicated: boolean;
  removed_at: string | null;
  removed_by: string | null;
  removal_reason: string | null;
  notes: string | null;
  created_at: string;
}

export interface CreateIndwellingDeviceRequest {
  patient_id: string;
  admission_id?: string;
  device_type: IndwellingDeviceType;
  site?: string;
  indication: string;
}

export interface ReviewIndwellingDeviceRequest {
  still_indicated: boolean;
  notes?: string;
}

export interface RemoveIndwellingDeviceRequest {
  removal_reason: string;
}

export interface InfectionDeviceDay {
  id: string;
  tenant_id: string;
  location_id: string;
  department_id?: string;
  record_date: string;
  patient_days: number;
  central_line_days: number;
  urinary_catheter_days: number;
  ventilator_days: number;
  recorded_by: string;
  created_at: string;
  updated_at: string;
}

export interface AntibioticStewardshipRequest {
  id: string;
  tenant_id: string;
  patient_id: string;
  antibiotic_name: string;
  dose?: string;
  route?: string;
  frequency?: string;
  duration_days?: number;
  indication: string;
  culture_sent: boolean;
  culture_result?: string;
  request_status: AntibioticRequestStatusType;
  requested_by: string;
  requested_at: string;
  reviewed_by?: string;
  reviewed_at?: string;
  review_notes?: string;
  escalation_reason?: string;
  auto_stop_date?: string;
  timeout_decision: string | null;
  timeout_decision_at: string | null;
  timeout_reviewed_by: string | null;
  timeout_notes: string | null;
  created_at: string;
  updated_at: string;
}

export interface AntibioticTimeoutReviewBody {
  decision: string;
  notes?: string;
}

export interface AntibioticConsumptionRecord {
  id: string;
  tenant_id: string;
  department_id?: string;
  antibiotic_name: string;
  atc_code?: string;
  record_month: string;
  quantity_used: number;
  ddd?: number;
  patient_days: number;
  ddd_per_1000_patient_days?: number;
  created_at: string;
  updated_at: string;
}

export interface BiowasteRecord {
  id: string;
  tenant_id: string;
  department_id: string;
  waste_category: WasteCategoryType;
  weight_kg: number;
  record_date: string;
  container_count: number;
  disposal_vendor?: string;
  manifest_number?: string;
  notes?: string;
  recorded_by: string;
  created_at: string;
  updated_at: string;
}

export interface NeedleStickIncident {
  id: string;
  tenant_id: string;
  incident_number: string;
  staff_id: string;
  incident_date: string;
  location_id?: string;
  department_id?: string;
  device_type: string;
  procedure_during?: string;
  body_part?: string;
  depth?: string;
  source_patient_id?: string;
  hiv_status?: string;
  hbv_status?: string;
  hcv_status?: string;
  pep_initiated: boolean;
  pep_details?: string;
  follow_up_schedule?: unknown;
  outcome?: string;
  reported_by: string;
  created_at: string;
  updated_at: string;
}

export interface HandHygieneAudit {
  id: string;
  tenant_id: string;
  audit_date: string;
  location_id?: string;
  department_id: string;
  auditor_id: string;
  observations: number;
  compliant: number;
  non_compliant: number;
  compliance_rate?: number;
  moment_breakdown?: unknown;
  staff_category?: string;
  findings?: string;
  created_at: string;
  updated_at: string;
}

export interface CultureSurveillance {
  id: string;
  tenant_id: string;
  culture_type: string;
  sample_site: string;
  location_id?: string;
  department_id?: string;
  collection_date: string;
  result?: string;
  organism?: string;
  colony_count?: number;
  acceptable?: boolean;
  action_taken?: string;
  collected_by: string;
  created_at: string;
  updated_at: string;
}

export interface OutbreakEvent {
  id: string;
  tenant_id: string;
  outbreak_number: string;
  organism: string;
  outbreak_status: OutbreakStatusType;
  detected_date: string;
  location_id?: string;
  department_id?: string;
  initial_cases: number;
  total_cases: number;
  description?: string;
  control_measures?: unknown;
  hicc_notified: boolean;
  hicc_notified_at?: string;
  containment_date?: string;
  closure_date?: string;
  root_cause?: string;
  lessons_learned?: string;
  reported_by: string;
  created_at: string;
  updated_at: string;
}

export interface OutbreakContact {
  id: string;
  tenant_id: string;
  outbreak_id: string;
  patient_id?: string;
  staff_id?: string;
  contact_type: string;
  exposure_date?: string;
  screening_date?: string;
  screening_result?: string;
  quarantine_required: boolean;
  quarantine_start?: string;
  quarantine_end?: string;
  notes?: string;
  created_at: string;
  updated_at: string;
}

// Request types
export interface CreateSurveillanceEventRequest {
  patient_id: string;
  hai_type: HaiType;
  organism?: string;
  device_type?: string;
  insertion_date?: string;
  infection_date: string;
  location_id?: string;
  department_id?: string;
  nhsn_criteria?: string;
  notes?: string;
}
export interface RecordDeviceDaysRequest {
  location_id: string;
  department_id?: string;
  record_date: string;
  patient_days: number;
  central_line_days: number;
  urinary_catheter_days: number;
  ventilator_days: number;
}
export interface CreateStewardshipRequest {
  patient_id: string;
  antibiotic_name: string;
  dose?: string;
  route?: string;
  frequency?: string;
  duration_days?: number;
  indication: string;
  culture_sent: boolean;
  escalation_reason?: string;
}
export interface ReviewStewardshipRequest {
  request_status: AntibioticRequestStatusType;
  review_notes?: string;
}
export interface CreateBiowasteRecordRequest {
  department_id: string;
  waste_category: WasteCategoryType;
  weight_kg: number;
  record_date: string;
  container_count: number;
  disposal_vendor?: string;
  manifest_number?: string;
  notes?: string;
}
export interface CreateNeedleStickIncidentRequest {
  staff_id: string;
  incident_date: string;
  location_id?: string;
  department_id?: string;
  device_type: string;
  procedure_during?: string;
  body_part?: string;
  depth?: string;
  source_patient_id?: string;
  pep_initiated: boolean;
  pep_details?: string;
}
export interface CreateHygieneAuditRequest {
  audit_date: string;
  department_id: string;
  observations: number;
  compliant: number;
  non_compliant: number;
  moment_breakdown?: unknown;
  staff_category?: string;
  findings?: string;
}
export interface CreateCultureSurveillanceRequest {
  culture_type: string;
  sample_site: string;
  location_id?: string;
  department_id?: string;
  collection_date: string;
  result?: string;
  organism?: string;
  colony_count?: number;
  acceptable?: boolean;
  action_taken?: string;
}
export interface CreateOutbreakRequest {
  organism: string;
  detected_date: string;
  location_id?: string;
  department_id?: string;
  initial_cases: number;
  description?: string;
}
export interface UpdateOutbreakRequest {
  outbreak_status?: OutbreakStatusType;
  total_cases?: number;
  control_measures?: unknown;
  hicc_notified?: boolean;
  root_cause?: string;
  lessons_learned?: string;
}
export interface CreateOutbreakContactRequest {
  patient_id?: string;
  staff_id?: string;
  contact_type: string;
  exposure_date?: string;
  quarantine_required: boolean;
  notes?: string;
}
